administrator/components/com_users/src/Table/TfaTable.php

```diff
@@ -0,0 +1,458 @@
+	 *
+	 * @return  boolean  True on success.
+	 *
+	 * @since __DEPLOY_VERSION__
+	 * @throws  \UnexpectedValueException
+	 */
+	public function delete($pk = null)
+	{
+		$record = $this;
+
+		if ($pk != $this->id)
+		{
+			$record = clone $this;
+			$record->reset();
+			$result = $record->load($pk);
+
+			if (!$result)
+			{
+				// If the record does not exist I will stomp my feet and deny your request
+				throw new RuntimeException(Text::_('JERROR_ALERTNOAUTHOR'), 403);
+			}
+		}
+
+		$user = Factory::getApplication()->getIdentity()
+			?? Factory::getContainer()->get(UserFactoryInterface::class)->loadUserById(0);
+
+		// The user must be a registered user, not a guest
+		if ($user->guest)
+		{
+			throw new RuntimeException(Text::_('JERROR_ALERTNOAUTHOR'), 403);
+		}
+
+		// You can only delete your own records, unless you're a super user or have delete privileges on this component
+		// phpcs:ignore
+		if (($record->user_id != $user->id) && !TfaHelper::canEditUser($user))
+		{
+			throw new RuntimeException(Text::_('JERROR_ALERTNOAUTHOR'), 403);
+		}
+
+		// Save flags used onAfterDelete
+		$this->deleteFlags[$record->id] = [
+			'default'    => $record->default,
+			// phpcs:ignore
+			'numRecords' => $this->getNumRecords($record->user_id),
+			// phpcs:ignore
+			'user_id'    => $record->user_id,
+			'method'     => $record->method,
+		];
+
+		$isDeleted = parent::delete($pk);
+
+		if ($isDeleted)
+		{
+			$this->afterDelete($pk);
+		}
+
+		return $isDeleted;
+	}
+
+	/**
+	 * Decrypt the possibly encrypted options
+	 *
+	 * @return void
+	 * @since __DEPLOY_VERSION__
+	 */
+	private function decryptOptions(): void
+	{
+		// Try with modern decryption
+		$decrypted = @json_decode($this->encryptService->decrypt($this->options ?? ''), true);
+
+		if (is_string($decrypted))
+		{
+			$decrypted = @json_decode($decrypted, true);
+		}
+
+		// FAll back to legacy decryption
+		if (!is_array($decrypted))
+		{
+			$decrypted = @json_decode($this->encryptService->decrypt($this->options ?? '', true), true);
+
+			if (is_string($decrypted))
+			{
+				$decrypted = @json_decode($decrypted, true);
+			}
+		}
+
+		$this->options = $decrypted ?: [];
+	}
+
+	/**
+	 * If this record is set to be the default, unset the default flag from the other records for the same user.
+	 *
+	 * @return void
+	 * @since __DEPLOY_VERSION__
+	 */
+	private function switchDefaultRecord(): void
+	{
+		if (!$this->default)
+		{
+			return;
+		}
+
+		/**
+		 * This record is marked as default, therefore we need to unset the default flag from all other records for this
+		 * user.
+		 */
+		$db    = $this->getDbo();
+		$zero  = 0;
+		$query = $db->getQuery(true)
+			->update($db->quoteName('#__user_tfa'))
+			->set($db->quoteName('default') . ' = :zero')
+			->where($db->quoteName('user_id') . ' = :user_id')
+			->where($db->quoteName('id') . ' != :id')
+			->bind(':zero', $zero, ParameterType::INTEGER)
```

Comment:
```suggestion
		$query = $db->getQuery(true)
			->update($db->quoteName('#__user_tfa'))
			->set($db->quoteName('default') . ' = 0')
			->where($db->quoteName('user_id') . ' = :user_id')
			->where($db->quoteName('id') . ' != :id')
```
It does not need bind variables for values which are constant and so can't be manipulated.

Reply:
Thanks! Will merge manually.